File: Capítulo_14/Programação dinâmica/Multiplicação de cadeia de matrizes/Códigos em Python/matrixChainOrder.py
# Algoritmo de multiplicação de cadeia de matrizes matrix_chain_order
def matrix_chain_order(p):
    n = len(p)
    m = [[0 for _ in range(n + 1)] for _ in range(n + 1)] # Inicialização da matrix m

    for l in range(2, n):
        for i in range(1, (n - l) + 1):
            j = (i + l) - 1
            m[i][j] = float("inf")

            for k in range(i, j):
                q =  m[i][k] + m[k + 1][j] + ((p[i - 1] * p[k]) * p[j])
                if q < m[i][j]:
                    m[i][j] = q

    return m[1][n - 1]

# Testando o algoritmo matrixChainOrder
p = [10, 100, 5, 50, 1]
print(matrix_chain_order(p)) # 1750

File: Capítulo_14/Programação dinâmica/Multiplicação de cadeia de matrizes/matrixChainOrder.js
// Algoritmo de multiplicação de cadeia de matrizes matrixChainOrder:
function matrixChainOrder(p) {
    const n = p.length;
    const m = [];
    for (let i = 0; i <= n; i++) {
        m[i] = [];
        m[i][i] = 0;
    }
    for (let l = 2; l < n; l++) {
        for (let i = 1; i <= (n - l) + 1; i++) {
            const j = (i + l) - 1;
            m[i][j] = Number.MAX_SAFE_INTEGER;
            for (let k = i; k <= j - 1; k++) {
                const q = m[i][k] + m[k + 1][j] + ((p[i - 1] * p[k]) * p[j]);
                if (q < m[i][j]) {
                    m[i][j] = q;
                }
            }
        }
    }
    return m[1][n - 1];
}

// Testando o algoritmo matrixChainOrder
const p = [10, 100, 5, 50, 1];
console.log(matrixChainOrder(p)); // 1750
